lib/rubocop/cop/style/legacy_hash.rb

```diff
@@ -0,0 +1,74 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks the usage of pre-2.4 `Hash[args]` method of converting arrays to hashes
+      #
+      # @example
+      #   # bad
+      #   Hash[ary]
+      #
+      #   # good
+      #   ary.to_h
+      #
+      #   # bad
+      #   Hash[key1, value1, key2, value2]
+      #
+      #   # good
+      #   {key1 => value1, key2 => value2}
+      class LegacyHash < Base
+        extend AutoCorrector
+
+        MSG_TO_H = 'Prefer ary.to_h to Hash[ary].'
+        MSG_LITERAL = 'Prefer literal hash to Hash[arg1, arg2, ...].'
+        MSG_SPLAT = 'Prefer array_of_pairs.to_h to Hash[*array].'
+        RESTRICT_ON_SEND = %i[[]].freeze
+
+        def_node_matcher :hash_from_array?, '(send (const nil? :Hash) :[] ...)'
+
+        def on_send(node)
+          return unless hash_from_array?(node)
+
+          # There are several cases:
+          # If there is one argument:
+          #   Hash[ary] => ary.to_h
+          #   Hash[*ary] => don't suggest corrections
+          # If there is 0 or 2+ arguments:
+          #   Hash[a1, a2, a3, a4] => {a1 => a2, a3 => a4}
+          #   ...but don't suggest correction if there is odd numbe of them (it is a bug)
+          node.arguments.count == 1 ? single_argument(node) : multi_argument(node)
+        end
+
+        private
+
+        def single_argument(node)
+          if node.arguments.first.splat_type?
+            add_offense(node, message: MSG_SPLAT)
```

Comment:
We should auto-correct this too.

Reply:
I don't think so. With splat, there are basically two possibilities:
1. Code inside `Hash[]` produces a flat list of keys and values by design (say, it reads a file which has `[key, value, key, value]` split line-by-line)
2. Code actually produces an array of pairs, and then `flatten(1)` added to the end of it to match `Hash[]` signature.

In both cases, it should be reviewed and redecided by human. 
The only working suggestion we can produce is
```diff
- Hash[*something]
+ something.each_slice(2).to_h
```
...and in most cases, it would not be a good suggestion.